Add the ability to customize the tile cache size

jczaplew opened the pull request:
This pull request addresses #4052. Currently the tile cache size is determined by the size of the viewport and there is no way for the user to manually override this size. This pull requests exposes a new option on the map, `cacheSize`, which can be used to explicitly set the tile cache size.

anandthakker commented:
Thanks for this PR< @jczaplew !

While do I agree that providing some control over the cache size seems like a good idea (especially in light of the iOS crashes), I find myself hesitant to solve this issue by exposing SourceCache's cache size as a direct option.

I'm going to spend a little time thinking about possible alternative designs and will post back here soon.

jczaplew commented:
Sounds good, and understandable!

mourner commented on src/ui/map.js:
```diff
@@ -57,7 +57,9 @@ const defaultOptions = {
-    refreshExpiredTiles: true
+    refreshExpiredTiles: true,
+
+    cacheSize: null
```
I would name it `tileCacheSize` for less ambiguity.

anandthakker commented on src/source/source_cache.js:
```diff
@@ -291,7 +294,10 @@ class SourceCache extends Evented {
         const commonZoomRange = 5;
-        this._cache.setMaxSize(Math.floor(approxTilesInView * commonZoomRange));
+
+        const maxSize = this._tileCacheSize || Math.floor(approxTilesInView * commonZoomRange);
```
Could we change this logic to:
```js
const viewDependentMaxSize = Math.floor(approxTilesInView * commonZoomRange);
const maxSize = typeof this._tileCacheSize === 'number' ? Math.min(this._tileCacheSize, viewDependentMaxSize) : viewDependentMaxSize;
```

This way, the new option would be setting an upper bound on the cache size, rather than forcing it to be fixed.

anandthakker commented on src/ui/map.js:
```diff
@@ -122,6 +124,7 @@ const defaultOptions = {
  * @param {number} [options.bearing=0] The initial bearing (rotation) of the map, measured in degrees counter-clockwise from north. If `bearing` is not specified in the constructor options, Mapbox GL JS will look for it in the map's style object. If it is not specified in the style, either, it will default to `0`.
  * @param {number} [options.pitch=0] The initial pitch (tilt) of the map, measured in degrees away from the plane of the screen (0-60). If `pitch` is not specified in the constructor options, Mapbox GL JS will look for it in the map's style object. If it is not specified in the style, either, it will default to `0`.
  * @param {boolean} [options.renderWorldCopies=true]  If `true`, multiple copies of the world will be rendered, when zoomed out.
+  * @param {number} [options.tileCacheSize=null]  The number of tiles stored in the tile cache. If omitted, the cache will be dynamically sized based on the current viewport.
```
In light of the comment on the logic in `SourceCache`, would it make sense to rename this to `maxTileCacheSize`?  (Sorry, I know you already renamed once! 😬 )

jczaplew replied:
Yes, I like that better in light of the above change

jczaplew commented:
@anandthakker - I renamed all instances of `tileCacheSize` to `maxTileCacheSize` per your recommendation, and also updated the logic in `src/source/source_cache.js` so that the parameter functions as a maximum value instead of simply a fixed value.

anandthakker commented:
Thanks @jczaplew!  Looks like there's a lint error preventing CI from passing.

anandthakker commented:
@jczaplew hm looks like some unit tests are failing.  Could you check that out?  You should be able to run them locally with `yarn test` or, slightly more conveniently `tap -b test/` to have it stop after the first failure.

jczaplew commented:
Unit tests check out. Please let me know if there are any other tweaks you would like.

anandthakker commented on src/ui/map.js:
```diff
@@ -122,6 +124,7 @@ const defaultOptions = {
  * @param {number} [options.bearing=0] The initial bearing (rotation) of the map, measured in degrees counter-clockwise from north. If `bearing` is not specified in the constructor options, Mapbox GL JS will look for it in the map's style object. If it is not specified in the style, either, it will default to `0`.
  * @param {number} [options.pitch=0] The initial pitch (tilt) of the map, measured in degrees away from the plane of the screen (0-60). If `pitch` is not specified in the constructor options, Mapbox GL JS will look for it in the map's style object. If it is not specified in the style, either, it will default to `0`.
  * @param {boolean} [options.renderWorldCopies=true]  If `true`, multiple copies of the world will be rendered, when zoomed out.
+  * @param {number} [options.maxTileCacheSize=null]  The maxiumum number of tiles stored in the tile cache. If omitted, the cache will be dynamically sized based on the current viewport.
```
Ah, just noticed this -- we should probably specify here that it's the maximum number of tiles in the cache **_for a given source_**

jczaplew replied:
Good catch. Fixed.

anandthakker commented:
Thanks @jczaplew 🎉 !

volkergraf commented:
Hmm I tried the following; (mapbox GL JS  0.39.1)

```
  map = new mapboxgl.Map({
            container: 'map',
            ......
            maxBounds: [
                [8.9651, 48.691],
                [9.0089, 48.705]
            ], // approximate bounds of tiles
            style: mapStyle,
            maxTileCacheSize: 2
           ....
        });
```

But when panning/tilting the whole (Cordova App) crashes and restarts .. any Ideas what I'm doing wrong ?

anandthakker commented:
@volkergraf I'm unable to reproduce this with the information above.  If you're able to create a minimal test case reproducing this issue, could you please open a new issue and share it there?
